chronos/ratelimit.nim

```diff
@@ -0,0 +1,136 @@
+#               Chronos Rate Limiter
+#            (c) Copyright 2022-Present
+#         Status Research & Development GmbH
+#
+#              Licensed under either of
+#  Apache License, version 2.0, (LICENSE-APACHEv2)
+#              MIT license (LICENSE-MIT)
+
+{.push raises: [Defect].}
+
+import ../chronos
+import timer
+
+export timer
+
+type
+  BucketWaiter = object
+    future: Future[void]
+    value: int
+    alreadyConsumed: int
+
+  TokenBucket* = ref object
+    budget: int
+    budgetCap: int
+    lastUpdate: Moment
+    fillDuration: Duration
+    workFuture: Future[void]
+    pendingRequests: seq[BucketWaiter]
+    manuallyReplenished: AsyncEvent
+
+proc update(bucket: TokenBucket) =
+  if bucket.fillDuration.milliseconds == 0:
+    bucket.budget = min(bucket.budgetCap, bucket.budget)
+    return
+
+  let
+    currentTime = Moment.now()
+    timeDelta = currentTime - bucket.lastUpdate
+    fillPercent = timeDelta.milliseconds.float / bucket.fillDuration.milliseconds.float
```

Comment:
if we use nanosecond precision here, we can simply assign the lastUpdate - there should be enough precision in the float, allowing `deltaFromReplenished` to be removed

Reply:
Not really, see the "Very long replenish" test
To fill by 1, we must wait 240ms, so if `update` is called before 240ms, it should not update `lastUpdate`, otherwise it may not refill properly
We could just `if replenished == 0: return`, but then it would be very imprecise for slow replenish / small buckets

Reply:
aaah, good point